indigo_app/static/javascript/indigo/views/document_editor.js

```diff
@@ -667,6 +672,40 @@
       this.textEditor.focus();
     },
 
+    getAnnotationAmendingWork: function(document) {
+      var date = document.get('expression_date'),
+          documentAmendments = document.get('amendments'),
+          amendments = [];
+
+      documentAmendments.forEach(function(amendment) {
+        if (amendment.date === date) {
+          amendments.push(amendment);
+        }
+      });
+
+      if (amendments) {
+        return amendments[0];
+      }
+
+    },
+
+    insertAnnotation: function(e) {
+      e.preventDefault();
+      var element = 'section/subsection/paragraph/subparagraph/<other> ()',
+          verb = 'substituted/amended/inserted/deleted/repealed/<other>',
+          amendingSection = 'section <insert>',
+          annotation;
+
+      if (this.annotationGenerator && this.annotationAmendingWork) {
+        annotation = this.annotationGenerator(element, verb, amendingSection, this.annotationAmendingWork);
```

Comment:
It feels to me that the generator should generate all of the text, given the various context elements. eg, provide it with:

* current document object
* amending work object
* verb (if chosen from menu, otherwise null)
* current element being edited (if any)

Then the generator can make use of those things to do something smart, including taking into account the document language. So all hardcoded strings are removed from this code.

Reply:
That would be very cool. I wanted to ask you about a (hopefully simple?) way of figuring out where we currently are in the document based on the information to hand.

In case it's relevant, just mentioning it won't be the amending work object but rather a serialization of it, i.e. its title and FRBR URI. Unless we want more, in which case I assume we'd have to change the serializer?

Reply:
> That would be very cool. I wanted to ask you about a (hopefully simple?) way of figuring out where we currently are in the document based on the information to hand.

`this.fragment`

> In case it's relevant, just mentioning it won't be the amending work object but rather a serialization of it, i.e. its title and FRBR URI. Unless we want more, in which case I assume we'd have to change the serializer?

That's fine for now.

Reply:
`this.fragment` only gives down to the section number, so we can't use it to generate e.g. `paragraph (b) substituted by …`

Reply:
Yeah, it's the best we have right now (could also be the whole document!) Useful to pass to the generator, though. It could make the assumption that the currently edited node is what changed, and the user could correct.

AFAIK, legislation.gov.uk only make amendment remarks down to the section level.

Reply:
It would be cool to use the amending work's `numbered_title` in the annotation (for e.g. `Act 3 of 2019`, `GN 45 of 2002`, and for localisation), but we'd have to inject that into a document's `amendment_events`, for which we'd have to change cobalt's `AmendmentEvent`. I think it might be worth doing, especially once we start working in multiple languages and/or with more provincial and national legislation.

Reply:
Simplest is to serialize the complete amending work information into the page as it loads, like we do with other expressions, and then look it up based on frbr uri.

Reply:
Oh, and then include `numbered_title` in the serialisation, which is actually useful even externally.